### PR TITLE
Hapi support

### PR DESCRIPTION
**Pull request by ranrib**
hapi

### CODE REVIEW

File: src/tracer.js [hunk 1]
```diff
@@ -203,7 +203,8 @@ module.exports.postTrace = function postTrace(traceObject) {
     utils.debugLog(`trace: ${JSON.stringify(traceObject, null, 2)}`);
     return session.post(
         config.getConfig().traceCollectorURL,
-        traceObject
+        traceObject,
+        { headers: { Authorization: `Bearer ${config.getConfig().token}` } }
```
Comment (ronyis) [inline src/tracer.js:207]: That's for future compatibility?
Isn't there a more suitable way to add authorization info without explicitly changing the HTTP headers?

File: src/wrappers/express.js [hunk 2]
```diff
@@ -28,6 +28,8 @@ function expressMiddleware(req, res, next) {
         tracer.addRunner(expressEvent);
     } catch (err) {
         utils.debugLog(err);
+        next();
```
Comment (ronyis) [inline src/wrappers/express.js:31]: Looks good...

File: src/wrappers/hapi.js [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+    // Handle response
+    response.then(() => {
+        try {
+            hapiRunner.finishRunner(hapiEvent, request, response, startTime);
+        } catch (err) {
+            tracer.addException(err);
+        }
+        tracer.sendTrace(() => {});
+    }).catch((err) => {
```
Comment (ronyis) [inline src/wrappers/hapi.js:72]: The code duplication here is just screaming :(

File: src/wrappers/hapi.js [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+
+module.exports = {
+    /**
+     * Initializes the Hapi tracer
+     */
+    init() {
+        if (Hapi && Hapi.server) {
+            shimmer.wrap(Hapi, 'server', hapiServerWrapper);
```
Comment (ronyis) [inline src/wrappers/hapi.js:122]: You verified that there isn't a builtin tracing I guess?

File: src/trace_context.js [hunk 5]
```diff
@@ -10,18 +10,31 @@ let namespace;
 
 /**
  * Creates an active context for tracer and run the handle
- * @param {Object} tracer object
+ * @param {Function} createTracer create a tracer object
  * @param {Function} handle function to run the context in
- * @param {Function} The runner event
  */
-function RunInContext(tracer, handle) {
+function RunInContext(createTracer, handle) {
```
Comment (ronyis) [inline src/trace_context.js]: Does it matter if this function will always return the value from `handle` function? It seems redundant to have two functions

**Review by ronyis — CHANGES_REQUESTED** [review]

File: src/tracer.js (hunk 1 above)
Comment (ranrib) [reply src/tracer.js:207]: Yes, that's for the future.
Updating the headers is the most common way for this

File: src/wrappers/express.js (hunk 2 above)
Comment (ranrib) [reply src/wrappers/express.js:31]: 👍

File: src/wrappers/hapi.js (hunk 3 above)
Comment (ranrib) [reply src/wrappers/hapi.js:72]: 👍

File: src/wrappers/hapi.js (hunk 4 above)
Comment (ranrib) [reply src/wrappers/hapi.js:122]: Indeed

File: src/trace_context.js (hunk 5 above)
Comment (ranrib) [reply src/trace_context.js]: You are right :)

**Comment by ranrib** [comment]
Ready for you @ronyis

File: test/acceptance/run.sh [hunk 6]
```diff
@@ -13,21 +13,22 @@ trap clenaup 1 2 15
 function run_acceptance_test() {
     build_num=$1
     runtime=$2
+    runtimeName=$3
     echo "deploying of ${runtime} [build: ${build_num}]"
     cp -r test/acceptance/ /tmp/acceptance/
     cd /tmp/acceptance/lambda-handlers/
     npm install $INITIAL_PWD
-    serverless deploy --runtime ${runtime} --buildNumber ${build_num} || {  echo "deployment of ${runtime} [build: ${build_num}] failed" ; result=1; }
+    serverless deploy --runtime ${runtime} --runtimeName ${runtimeName} --buildNumber ${build_num} || {  echo "deployment of ${runtime} [build: ${build_num}] failed" ; result=1; }
```
Comment (ronyis) [inline test/acceptance/run.sh:21]: Looks like using the stage to include build number + runtime can make things easier

**Review by ronyis — APPROVED** [review]

**Pull request merged**